File: sdk/typescript/src/launcher.ts
import type { AclipApp } from "./app.js";

export type CliAppTarget =
  | AclipApp
  | (() => AclipApp)
  | (() => Promise<AclipApp>);

export async function resolveApp(target: CliAppTarget): Promise<AclipApp> {
  if (typeof target === "function") {
    return await target();
  }
  return target;
}

export async function cliMain(
  target: CliAppTarget,
  argv: string[] = process.argv.slice(2)
): Promise<number> {
  const exitCode = await (await resolveApp(target)).run(argv);
  process.exitCode = exitCode;
  return exitCode;
}

File: sdk/python/src/aclip/launcher.py
from __future__ import annotations

from typing import Callable

from .app import AclipApp

AppTarget = AclipApp | Callable[[], AclipApp] | str


def resolve_app(target: AppTarget) -> AclipApp:
    if isinstance(target, AclipApp):
        return target

    if isinstance(target, str):
        from .packaging import load_app_factory

        return load_app_factory(target)()

    return target()


def cli_main(target: AppTarget, argv: list[str] | None = None) -> None:
    raise SystemExit(resolve_app(target).run(argv))
